packages/sw-cacheable-response-behavior/src/lib/behavior.js

```diff
@@ -0,0 +1,121 @@
+/*
+ Copyright 2016 Google Inc. All Rights Reserved.
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+import assert from '../../../../lib/assert';
+
+/**
+ * @memberof module:sw-cacheable-response-behavior
+ *
+ * @example
+ * // Used as an automatically invoked as "behavior" by a RequestWrapper:
+ *
+ * const cacheName = 'runtime-cache';
+ * const requestWrapper = new goog.runtimeCaching.RequestWrapper({
+ *   cacheName,
+ *   behaviors: [
+ *     new goog.cacheableResponse.Behavior({
+ *       statuses: [0, 200, 404]
+ *     })
+ *   ]
+ * });
+ *
+ * // Set up a route to match any requests made against the example.com domain.
+ * // The responses will be cached if the response code is 0, 200, or 404, and
+ * // will not be cached otherwise.
+ * const route = new goog.routing.RegExpRoute({
+ *   match: ({url}) => url.domain === 'example.com',
+ *   handler: new goog.runtimeCaching.StaleWhileRevalidate({requestWrapper})
+ * });
+ *
+ * @example
+ * // Explicitly invoked usage independent of the goog.routing framework, via
+ * // the isResponseCacheable() method:
+ *
+ * // TODO: Write sample code.
+ */
+class Behavior {
+  /**
+   * Creates a new `Behavior` instance, which stores configuration and logic
+   * to determine whether a `Response` object is cacheable or not.
+   *
+   * If multiple criteria are present (e.g. both `statuses` and `headers`), then
+   * the `Response` needs to meet all of the criteria to be cacheable.
+   *
+   * @param {Object} input The input object to this function.
+   * @param {Array<Number>} [input.statuses] The status codes that are
+   *        checked when determining whether a `Response` is cacheable.
+   * @param {Object<String,String>} [input.headers] The header values that are
+   *        checked when determining whether a `Response` is cacheable.
+   */
+  constructor({statuses, headers} = {}) {
```

Comment:
Was there a reason this only takes statuses and headers?

It feels like a callback with the response would cater for a lot more scenarios than enforcing only these options.

For example, if the developer wanted to only cache responses with a certain set of headers '.some' wouldn't be correct here. I'm also wondering if the developer would ever want to restrict caching based on the request as well as the response (i.e. only cache is the request header and the response header match....Edge case I know)